python/ray/data/_internal/compute.py

```diff
@@ -254,25 +281,35 @@ def ready(self):
 
             def map_block_split(
                 self,
-                block: Block,
                 input_files: List[str],
-                *fn_args,
+                num_blocks: int,
+                *blocks_and_fn_args,
                 **fn_kwargs,
             ) -> BlockPartition:
                 return _map_block_split(
-                    block, block_fn, input_files, self.fn, *fn_args, **fn_kwargs
+                    block_fn,
+                    input_files,
+                    self.fn,
+                    num_blocks,
+                    *blocks_and_fn_args,
+                    **fn_kwargs,
                 )
 
             @ray.method(num_returns=2)
             def map_block_nosplit(
                 self,
-                block: Block,
                 input_files: List[str],
-                *fn_args,
+                num_blocks: int,
+                *blocks_and_fn_args,
```

Comment:
curious why not pass `blocks: List[Block]` as parameter separately?

Reply:
The blocks need to be top-level arguments to the task function in order for the task dependency resolution to happen; if we passed it as a list of block references, the task would be scheduled without waiting for those blocks to be materialized, we'd have to manually `ray.get()` them, data locality wouldn't be taken into account when scheduling the tasks, etc.